File: solutions/0743. Network Delay Time/0743.cpp
struct T {
  int u;
  int d;
  T(int u, int d) : u(u), d(d) {}
};

class Solution {
 public:
  int networkDelayTime(vector<vector<int>>& times, int n, int k) {
    vector<vector<pair<int, int>>> graph(n);
    auto compare = [](const T& a, const T& b) { return a.d > b.d; };
    priority_queue<T, vector<T>, decltype(compare)> minHeap(compare);
    vector<bool> seen(n);

    for (const auto& t : times) {
      const int u = t[0] - 1;
      const int v = t[1] - 1;
      const int w = t[2];
      graph[u].emplace_back(v, w);
    }

    minHeap.emplace(k - 1, 0);

    while (!minHeap.empty()) {
      const auto [u, d] = minHeap.top();
      minHeap.pop();
      if (seen[u])
        continue;
      seen[u] = true;
      if (--n == 0)
        return d;
      for (const auto& [v, w] : graph[u])
        minHeap.emplace(v, d + w);
    }

    return -1;
  }
};

File: solutions/0743. Network Delay Time/0743.java
class T {
  public int u;
  public int d;
  public T(int u, int d) {
    this.u = u;
    this.d = d;
  }
}

class Solution {
  public int networkDelayTime(int[][] times, int n, int k) {
    List<Pair<Integer, Integer>>[] graph = new List[n];
    PriorityQueue<T> minHeap = new PriorityQueue<>((a, b) -> a.d - b.d);
    boolean[] seen = new boolean[n];

    for (int i = 0; i < n; ++i)
      graph[i] = new ArrayList<>();

    for (int[] t : times) {
      final int u = t[0] - 1;
      final int v = t[1] - 1;
      final int w = t[2];
      graph[u].add(new Pair<>(v, w));
    }

    minHeap.offer(new T(k - 1, 0));

    while (!minHeap.isEmpty()) {
      final int u = minHeap.peek().u;
      final int d = minHeap.poll().d;
      if (seen[u])
        continue;
      seen[u] = true;
      if (--n == 0)
        return d;
      for (var node : graph[u]) {
        final int v = node.getKey();
        final int w = node.getValue();
        minHeap.offer(new T(v, d + w));
      }
    }

    return -1;
  }
}
